doc/source/whatsnew/v1.2.0.rst

```diff
@@ -620,6 +620,7 @@ Indexing
 - Bug in indexing on a :class:`Series` or :class:`DataFrame` with a :class:`CategoricalIndex` using listlike indexer that contains elements that are in the index's ``categories`` but not in the index itself failing to raise ``KeyError`` (:issue:`37901`)
 - Bug in :meth:`DataFrame.iloc` and :meth:`Series.iloc` aligning objects in ``__setitem__`` (:issue:`22046`)
 - Bug in :meth:`DataFrame.loc` did not raise ``KeyError`` when missing combination was given with ``slice(None)`` for remaining levels (:issue:`19556`)
+- Bug in :meth:`MultiIndex.remove_unused_levels` was dropping missing values when levels contain ``NaN`` set by ``set_levels`` (:issue:`37510`)
```

Comment:
can you make this more user facing, e.g. after an unstack with missing levels the index names are not correct (the fact that remove_usused_levels is called is not really important here).